### PR TITLE
Create ALLOW_X_SKINS game info flag

### PR DESCRIPTION
**Pull request by def-**
To allow server to set any x_ prefixed skin and client won't filter it
out. As requested by Pure_luck for his mod to show players as tanks,
walls, etc. Won't be enabled on DDNet-Servers, thus such skins can be
added where a server modification wants to fine-control what skins are
allowed and can enforce such skins.

### CODE REVIEW

**Comment by fokkonaut** [comment]
Well, can a client use such skin on their own then without it getting forced by the server?

File: src/game/server/gamecontroller.cpp [hunk 1]
```diff
@@ -539,24 +539,25 @@ void IGameController::Snap(int SnappingClient)
...
+		GAMEINFOFLAG_ALLOW_ZOOM |
+		GAMEINFOFLAG_BUG_DDRACE_GHOST |
+		GAMEINFOFLAG_BUG_DDRACE_INPUT |
+		GAMEINFOFLAG_PREDICT_DDRACE |
+		GAMEINFOFLAG_PREDICT_DDRACE_TILES |
+		GAMEINFOFLAG_ENTITIES_DDNET |
+		GAMEINFOFLAG_ENTITIES_DDRACE |
+		GAMEINFOFLAG_ENTITIES_RACE |
+		GAMEINFOFLAG_RACE;
+	pGameInfoEx->m_Flags2 = 0;
```
Comment (pure-luck-999) [inline src/game/server/gamecontroller.cpp:560]: Server owners should replace 0 with ALLOW_X_SKINS here, right?

Comment (fokkonaut) [reply src/game/server/gamecontroller.cpp:560]: Yes.

**Comment by def-** [comment]
> Well, can a client use such skin on their own then without it getting forced by the server?

Depends on how the server implements the filtering. But client can try setting it, depends on server if it gets respected. Won't have to be implemented in this project but in mods that want to use allow_x_skins.

**Comment by heinrich5991** [comment]
I'd like to review this one later.

File: src/game/client/gameclient.cpp [hunk 2]
```diff
@@ -1009,14 +1009,21 @@ static CGameInfo GetGameInfo(const CNetObj_GameInfoEx *pInfoEx, int InfoExSize,
 	{
 		Flags = pInfoEx->m_Flags;
 	}
+	int Flags2 = 0;
+	if(Version >= 5)
```
Comment (heinrich5991) [inline src/game/client/gameclient.cpp:1013]: Needs a size check as well. If it fails, just set version to 4 I guess.

File: src/game/client/gameclient.cpp [hunk 3]
```diff
@@ -1069,6 +1078,7 @@ static CGameInfo GetGameInfo(const CNetObj_GameInfoEx *pInfoEx, int InfoExSize,
 	Info.m_EntitiesBW = BlockWorlds;
 	Info.m_Race = Race;
 	Info.m_DontMaskEntities = !DDNet;
+	Info.m_AllowXSkins = AllowXSkins;
```
Comment (heinrich5991) [inline src/game/client/gameclient.cpp:1081]: This is a bit implicit.

AllowXSkins is intended to be false unless the server sends an version >= 5 info with the flag enabled.

Can you set `Info.m_AllowXSkins` to false here, and later set it to `Flags2 & GAMEINFOFLAG2_ALLOW_X_SKINS` in a `if(Version >= 5)` branch after `if(Version >= 4)`?

File: src/game/client/gameclient.cpp [hunk 4]
```diff
@@ -1025,19 +1032,21 @@ static CGameInfo GetGameInfo(const CNetObj_GameInfoEx *pInfoEx, int InfoExSize,
...
-		Plus = Flags&GAMEINFOFLAG_GAMETYPE_PLUS;
+		Race = Flags & GAMEINFOFLAG_GAMETYPE_RACE;
+		FastCap = Flags & GAMEINFOFLAG_GAMETYPE_FASTCAP;
+		FNG = Flags & GAMEINFOFLAG_GAMETYPE_FNG;
+		DDRace = Flags & GAMEINFOFLAG_GAMETYPE_DDRACE;
+		DDNet = Flags & GAMEINFOFLAG_GAMETYPE_DDNET;
+		BlockWorlds = Flags & GAMEINFOFLAG_GAMETYPE_BLOCK_WORLDS;
+		Vanilla = Flags & GAMEINFOFLAG_GAMETYPE_VANILLA;
+		Plus = Flags & GAMEINFOFLAG_GAMETYPE_PLUS;
+		City = Flags2 & GAMEINFOFLAG2_GAMETYPE_CITY;
```
Comment (heinrich5991) [inline src/game/client/gameclient.cpp:1049]: This line should be in the a `if(Version >= 5)`. Alternatively: `Version >= 5 && (Flags2 & GAMEINFOFLAG2_GAMETYPE_CITY`.

**Review by heinrich5991 — CHANGES_REQUESTED** [review]
Needs a size check (otherwise we could read memory not intended for us), rest seems to be correct.

File: src/game/client/gameclient.cpp (hunk 2 above)
Comment (def-) [reply src/game/client/gameclient.cpp:1013]: Done

File: src/game/client/gameclient.cpp (hunk 3 above)
Comment (def-) [reply src/game/client/gameclient.cpp:1081]: Done

File: src/game/client/gameclient.cpp (hunk 4 above)
Comment (def-) [reply src/game/client/gameclient.cpp:1049]: But Flags2 is only assigned if Version >= 5, so not sure what good the check does.(except being more explicit)

File: src/game/client/gameclient.cpp [hunk 5]
```diff
@@ -996,10 +996,14 @@ void CGameClient::ProcessEvents()
...
 {
 	int Version = -1;
-	if(InfoExSize >= 8)
+	if(InfoExSize >= 12)
 	{
 		Version = pInfoEx->m_Version;
 	}
+	else if(InfoExSize >= 8)
+	{
+		Version = 4;
```
Comment (heinrich5991) [inline src/game/client/gameclient.cpp:1005]: This stops version 1-3 from working. Need to cap it at 4 instead. Maybe: `Version = min(pInfoEx->m_Version, 4);`

Comment (def-) [reply src/game/client/gameclient.cpp:1005]: Done

**Review by heinrich5991 — APPROVED** [review]

**Comment by heinrich5991** [comment]
bors r+

Thanks! :)

**Pull request merged**